### PR TITLE
Recognizing Jenkinsfile as Groovy source

### PR DESCRIPTION
**Pull request by jglick**
`Jenkinsfile` is the conventional name for a Groovy-syntax script containing build instructions for the Jenkins continuous integration server using the Pipeline feature. [Details](https://github.com/jenkinsci/workflow-plugin/blob/7d2004ea482418f84c3f221d3f55dff20e8ada44/TUTORIAL.md#creating-multibranch-projects) and [context](https://issues.jenkins-ci.org/browse/JENKINS-31152).

Look to be about a couple hundred [known files](https://github.com/search?utf8=%E2%9C%93&q=Jenkinsfile+in%3Apath&type=Code&ref=searchresults) out there. To date the trick for getting syntax highlighting has been to use `#!groovy` as [here](https://github.com/jenkinsci/pipeline-examples/blob/0a7aa6180e14fd17901a514c5e6bb356e416f41f/global-library-examples/global-function/Jenkinsfile#L1).

For my colleagues: @reviewbybees esp. @kohsuke who selected this filename.

### CODE REVIEW

File: lib/linguist/vendor.yml [hunk 1]
```diff
@@ -302,3 +302,6 @@
 
 # Android Google APIs
 - (^|/)\.google_apis/
+
+# Jenkins Pipeline
+- ^Jenkinsfile$
```
Comment (jglick) [inline lib/linguist/vendor.yml]: Typically a `Jenkinsfile` will be at the root of its repository; when used as part of a multibranch Pipeline project, that is required as this exact filename is the indication that a repository/branch is ready to build.

File: samples/Groovy/filenames/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
+jettyUrl = 'http://localhost:8081/'
```
Comment (jglick) [inline samples/Groovy/filenames/Jenkinsfile]: Adapted from a [standard demo](https://github.com/jenkinsci/workflow-plugin/blob/7d2004ea482418f84c3f221d3f55dff20e8ada44/demo/repo/Jenkinsfile).

File: test/test_file_blob.rb [hunk 3]
```diff
@@ -522,6 +522,9 @@ def test_vendored
     assert sample_blob("myapp/My Template.xctemplate/___FILEBASENAME___.h").vendored?
     assert sample_blob("myapp/My Images.xcassets/some/stuff.imageset/Contents.json").vendored?
     assert !sample_blob("myapp/MyData.json").vendored?
+
+    # Jenkins
+    assert sample_blob("Jenkinsfile").vendored?
```
Comment (jglick) [inline test/test_file_blob.rb]: By the way `lib/linguist/vendor.yml` suggests adding test coverage to `test/test_blob.rb#test_vendored` but it seems that almost all of the existing tests are here instead, so I tried to follow suit.

File: samples/Groovy/filenames/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+def servers
+
+stage 'Dev'
+node {
+    checkout scm
+    servers = load 'servers.groovy'
```
Comment (jglick) [inline samples/Groovy/filenames/Jenkinsfile]: Auxiliary library files are conventionally just named `*.groovy` and so require no special treatment.

**Comment by amuniz** [comment]
Nice. 
:bee: and :+1: !

**Comment by recena** [comment]
:bee: and :+1: undoubtedly!

File: lib/linguist/vendor.yml (hunk 1 above)
Comment (arfon) [inline lib/linguist/vendor.yml]: @jglick - is there a reason this should be considered a `vendored?` file for Linguist - this basically means that it won't contribute to language statistics in the GitHub repository (the coloured bar at the top of the repository page)

File: test/test_file_blob.rb (hunk 3 above)
Comment (arfon) [inline test/test_file_blob.rb]: :+1: yeah, this is fine.

**Comment by arfon** [comment]
Thanks for the very nice PR @jglick. I've left a couple of comment inline.

File: lib/linguist/vendor.yml (hunk 1 above)
Comment (jglick) [inline lib/linguist/vendor.yml]: Yes, that was intentional, and by analogy with very domain-specific files like `Vagrantfile` which you would not consider associated with Ruby development. Groovy is used as a host for the EDSL, but examples like [this](https://github.com/jenkinsci/pipeline-examples/blob/0a7aa6180e14fd17901a514c5e6bb356e416f41f/global-library-examples/global-function/Jenkinsfile#L5-L15) show that the actual `Jenkinsfile` can be made to look more like a config file than a program. Furthermore, allowable methods calls are restricted by a sandbox to a whitelist, and execution proceeds via a custom interpreter via continuation-passing transform which imposes various constraints on serializability etc., so the result does not look much like a general-purpose Groovy program.

**Comment by jglick** [comment]
@reviewbybees done

File: lib/linguist/vendor.yml (hunk 1 above)
Comment (arfon) [inline lib/linguist/vendor.yml]: OK great. Thanks for the clarification.

**Comment by arfon** [comment]
Thanks for the contribution @jglick! This will be live in the next release of Linguist (probably another week or so before we cut a new gem)

**Pull request merged**